Composer/packages/tools/language-servers/language-generation/src/LGServer.ts

```diff
@@ -0,0 +1,339 @@
+      };
+      completionList.push(item);
+    });
+
+    buildInfunctionsMap.forEach((value, key) => {
+      const item = {
+        label: key,
+        kind: CompletionItemKind.Function,
+        insertText: key + '(' + this.removeParamFormat(value.Params.toString()) + ')',
+        documentation: value.Introduction,
+      };
+      completionList.push(item);
+    });
+
+    const match = this.matchedStates(params);
+    // TODO: more precise match
+    if (match && match.matched && match.state && allowedCompletionStates.includes(match.state.toLowerCase())) {
+      return Promise.resolve({ isIncomplete: true, items: completionList });
+    } else {
+      return Promise.resolve(null);
+    }
+  }
+
+  protected validate(document: TextDocument): void {
+    this.cleanPendingValidation(document);
+    document.uri,
+      setTimeout(() => {
+        this.pendingValidationRequests.delete(document.uri);
+        this.doValidate(document);
+      });
+  }
+
+  protected cleanPendingValidation(document: TextDocument): void {
+    const request = this.pendingValidationRequests.get(document.uri);
+    if (request !== undefined) {
+      clearTimeout(request);
+      this.pendingValidationRequests.delete(document.uri);
+    }
+  }
+
+  protected doValidate(document: TextDocument): void {
+    let text = document.getText();
+    let lineOffset = 0;
+    const LGDocument = this.getLGDocument(document);
+    if (LGDocument) {
+      // concat new content for validate
+      const { content, template } = LGDocument;
+      const updatedTemplate = {
+        Name: template.Name,
+        Body: text,
+      };
+      text = updateTemplateInContent(content, updatedTemplate);
+      lineOffset = getTemplatePositionOffset(content, updatedTemplate);
+    }
+
+    if (text.length === 0) {
+      this.cleanDiagnostics(document);
+      return;
+    }
+    const staticChercher = new lg.StaticChecker();
+    const lgDiags = staticChercher.checkText(text, '', lg.ImportResolver.fileResolver);
+    const diagnostics: Diagnostic[] = [];
+    lgDiags.forEach(diag => {
+      const diagnostic: Diagnostic = {
+        severity: convertSeverity(diag.Severity),
+        range: {
+          start: Position.create(diag.Range.Start.Line - 1 - lineOffset, diag.Range.Start.Character),
+          end: Position.create(diag.Range.End.Line - 1 - lineOffset, diag.Range.End.Character),
+        },
+        message: diag.Message,
+        source: document.uri,
+      };
+      diagnostics.push(diagnostic);
+    });
+
+    this.sendDiagnostics(document, diagnostics);
+  }
+
+  protected cleanDiagnostics(document: TextDocument): void {
+    this.sendDiagnostics(document, []);
+  }
+
+  protected sendDiagnostics(document: TextDocument, diagnostics: Diagnostic[]): void {
+    this.connection.sendDiagnostics({
+      uri: document.uri,
+      diagnostics,
+    });
+  }
+}
+
+export function start(reader: MessageReader, writer: MessageWriter): LGServer {
```

Comment:
We don't have to export a `start()`

Let's create a simple interface 

ILanguageServer
{
  start();
}

that's it,decouple the creation with starting. let the caller decide when to start.